### PR TITLE
PHP: proto3 adoption

### LINKED ISSUE #6362: Deduplicate .proto files under src/php once PHP supports proto3.
Context: #4123

Once proto3 is supported, these files  should be removed and canonical versions from src/proto should be used.
```
src/php/tests/generated_code/math.proto
src/php/tests/interop/test.proto
src/php/tests/interop/empty.proto
src/php/tests/interop/messages.proto
```

### LINKED ISSUE #6637: PHP protoc plugin
Implement the gRPC PHP protoc plugin, after Proto3 PHP support has been rolled out.

### PR DESCRIPTION
**Pull request by stanley-cheung**
Fixes #6637 
Fixes #6362
Step towards #5983

 * This is not ready to be merged yet. We need `google/protobuf` to add the tag `v3.1.0-alpha-1` and release the package to Packagist/composer
 * Added the gRPC PHP Plugin to generate client service stub classes
 * Removed the previous `stanley-cheung/Protobuf-PHP` library and ruby dependency but if user decides to continue to use that library things will continue to work
 * Updated the Math generated code test, and interop tests
 * Removed the duplicated sources `math.proto`, `test.proto`, `empty.proto`, `messages.proto`. Now we can share the sources with `src/protos`
 * Updated jenkins PHP dockerfiles to use this new library, can now remove the ruby dependency.
 * Updated README
 * No change to the PECL extension

To test code generation, please run `src/php/bin/generate_proto_php.sh`.

From user's perspective, this is almost a drop-in replacement of the previous `Protobuf-PHP` library. See the changes in `src/php/tests/generated_code/AbstractGeneratedCodeTest.php`. Aside from a different code generator, all we need to change is to include the re-generated stub code, and the client code continues to work.

Next steps:
 * Update the `examples/php` directory helloworld greeter and route guide example. But that will need our `grpc/grpc` version 1.1 to be released to Packagist/composer first.
 * Update grpc.github.io website and change any reference to the old `Protobuf-PHP` library.

Tests won't work yet because the `composer install` step will fail because the `google/protobuf` package is not available yet.

### CODE REVIEW

**Comment by stanley-cheung** [comment]
@TeBoring FYI

**Comment by thinkerou** [comment]
But protobuf not support php7 now and compile error on mac.

File: BUILD [hunk 1]
```diff
@@ -2001,6 +2001,8 @@ cc_library(
     "src/compiler/node_generator_helpers.h",
     "src/compiler/objective_c_generator.h",
     "src/compiler/objective_c_generator_helpers.h",
+    "src/compiler/php_generator.h",
```
Comment (jtattermusch) [inline BUILD]: nit: in case there's a lot of changes to the generated files, it is better to put all the generated changes in a separate commit, because that makes reviewing easier.

File: src/php/README.md [hunk 2]
```diff
@@ -98,44 +128,48 @@ $ ./bin/run_tests.sh
...
 
-If you compiled the gRPC C core library from source above, the `protoc` binary should have been installed as well. If it hasn't been installed, you can run the following commands to install it.
+If `protoc` hasn't been installed, you can run the following commands to
+install it.
 
 ```sh
 $ cd grpc/third_party/protobuf
-$ sudo make install   # 'make' should have been run by core grpc
+$ ./autogen.sh && ./configure && make
+$ sudo make install
```
Comment (jtattermusch) [inline src/php/README.md]: In general we've put great effort into preventing the need to install protobuf by users (because they are almost guaranteed to run into issues once they need to upgrade protobuf one day).  For other languages, we provide grpc.tools packages that contain precompiled protoc and protoc plugin, so that users don't have to make install protobuf.

In this PR, I think is is fine to leave things, but expect that we will need to get rid of the make install soon.

File: src/php/README.md [hunk 3]
```diff
@@ -98,44 +128,48 @@ $ ./bin/run_tests.sh
...
 ```
 
-Alternatively, you can download `protoc` binaries from [the protocol buffers Github repository](https://github.com/google/protobuf/releases).
+Alternatively, you can download the `protoc` binaries from
```
Comment (jtattermusch) [inline src/php/README.md]: Ok, this should be the default and recommended action for the users - let's add warning not do make install protobuf  (or do it at your own risk).

File: src/php/README.md [hunk 4]
```diff
@@ -98,44 +128,48 @@ $ ./bin/run_tests.sh
...
 
 ```sh
-$ git clone https://github.com/stanley-cheung/Protobuf-PHP
-$ cd Protobuf-PHP
-$ gem install rake ronn
-$ rake pear:package version=1.0
-$ sudo pear install Protobuf-1.0.tgz
+$ git clone https://github.com/grpc/grpc
+$ make plugins
+$ sudo make install-plugins
```
Comment (jtattermusch) [inline src/php/README.md]: Same as above, eventually there should be a package that provides protoc and plugins (once you install here, there's no way of uninstalling and that's really bad).

File: src/php/bin/generate_proto_php.sh [hunk 5]
```diff
@@ -28,13 +28,32 @@
...
 
-
 set +e
-cd $(dirname $0)
+cd $(dirname $0)/../../..
+
+protoc --proto_path=src/proto/math \
+       --php_out=src/php/tests/generated_code \
+       --grpc_out=src/php/tests/generated_code \
+       --plugin=protoc-gen-grpc=$(which grpc_php_plugin) \
```
Comment (jtattermusch) [inline src/php/bin/generate_proto_php.sh]: assumes php plugin is installed - see above.

Comment (jtattermusch) [inline src/php/bin/generate_proto_php.sh]: Also below.

File: src/php/bin/generate_proto_php.sh [hunk 6]
```diff
@@ -28,13 +28,32 @@
...
+       src/proto/math/math.proto
 
-gen_code='../tests/generated_code'
-interop='../tests/interop'
+# replace the Empty message with EmptyMessage
+# because Empty is a PHP reserved word
```
Comment (jtattermusch) [inline src/php/bin/generate_proto_php.sh]: :-)

Comment (jtattermusch) [inline src/php/bin/generate_proto_php.sh]: Actually at some point the protoc for PHP should be clever enough and have a list of reserved words and alter the names in the generated code so they don't collide. (Empty becomes `Empty_` or so). Is there a plan for that?

File: src/php/bin/generate_proto_php.sh (hunk 5 above)
Comment (jtattermusch) [inline src/php/bin/generate_proto_php.sh]: Actually, to regenerate code in our repository (not in examples), you should use the plugins from 
bins/opt/grpc_php_plugins (they'll be there after make plugins) - otherwise you'll be generating code with a different version of protobuf than the one in third_party/protobuf (and that's the one that should be used).

File: src/compiler/php_plugin.cc [hunk 7]
```diff
@@ -0,0 +1,73 @@
+/*
+ *
+ * Copyright 2015, Google Inc.
```
Comment (thinkerou) [inline src/compiler/php_plugin.cc]: Should be 2016

File: src/php/composer.json [hunk 8]
```diff
@@ -8,7 +8,7 @@
   "version": "1.1.0",
   "require": {
     "php": ">=5.5.0",
-    "stanley-cheung/protobuf-php": "v0.6"
+    "stanley-cheung/protobuf": "v3.1.0-alpha-1"
```
Comment (thinkerou) [inline src/php/composer.json]: Should be `google/protobuf`?

File: src/php/lib/Grpc/AbstractCall.php [hunk 9]
```diff
@@ -121,6 +121,23 @@ public function cancel()
...
+     * @return string The protobuf binary format
+     */
+    protected function serializeMessage($data) {
+        // Proto3 implementation
+        if (method_exists($data, 'encode')) {
+            return $data->encode();
+        }
+
+        // Protobuf-PHP implementation
+        return $data->serialize();
```
Comment (thinkerou) [inline src/php/lib/Grpc/AbstractCall.php]: User needs the official `protobuf` and `protobuf-php`? So user need to install the two lib.

Comment (thinkerou) [inline src/php/lib/Grpc/AbstractCall.php]: Oh, in order to compatible `protobuf-php`, right?

File: templates/src/php/composer.json.template [hunk 10]
```diff
@@ -10,7 +10,7 @@
     "version": "${settings.php_version.php_composer()}",
     "require": {
       "php": ">=5.5.0",
-      "stanley-cheung/protobuf-php": "v0.6"
+      "stanley-cheung/protobuf": "v3.1.0-alpha-1"
```
Comment (thinkerou) [inline templates/src/php/composer.json.template]: `google/protobuf`?

File: src/php/bin/generate_proto_php.sh (hunk 5 above)
Comment (stanley-cheung) [inline src/php/bin/generate_proto_php.sh]: Yes, that's #5983

File: src/php/README.md (hunk 2 above)
Comment (stanley-cheung) [inline src/php/README.md]: Yes that's #5983

File: src/php/composer.json (hunk 8 above)
Comment (stanley-cheung) [inline src/php/composer.json]: Removed

File: templates/src/php/composer.json.template (hunk 10 above)
Comment (stanley-cheung) [inline templates/src/php/composer.json.template]: Removed

File: BUILD (hunk 1 above)
Comment (stanley-cheung) [inline BUILD]: I have split this into multiple smaller commits in #8179

File: src/php/README.md (hunk 3 above)
Comment (stanley-cheung) [inline src/php/README.md]: Added warning in #8179

File: src/php/README.md (hunk 4 above)
Comment (stanley-cheung) [inline src/php/README.md]: Added note that we will create better packages for the plugin in the future in #8179

File: src/php/bin/generate_proto_php.sh (hunk 5 above)
Comment (stanley-cheung) [inline src/php/bin/generate_proto_php.sh]: Fixed

File: src/php/bin/generate_proto_php.sh (hunk 6 above)
Comment (stanley-cheung) [inline src/php/bin/generate_proto_php.sh]: I have discussed with the protobuf team and that's the solution we settled for now. I will keep pushing for a more automatic way of dealing with Empty in the future.

File: src/compiler/php_plugin.cc (hunk 7 above)
Comment (stanley-cheung) [inline src/compiler/php_plugin.cc]: Fixed

File: src/php/lib/Grpc/AbstractCall.php (hunk 9 above)
Comment (stanley-cheung) [inline src/php/lib/Grpc/AbstractCall.php]: Yes just for backward-compatibility. Can remove in the future

**Comment by stanley-cheung** [comment]
Closing this in favor of #8179
